### PR TITLE
Build CoreFx with a mounted docker volume

### PR DESCRIPTION
**Pull request by chcosta**
For CoreFx Linux official builds, rather than starting a detached docker container and issuing commands to it; issue each docker command to a new docker container using a mounted volume.  This method means that we share the disk drive between the Host machine and docker containers which allows us to grab logs / binaries /etc... whenever and not have to deal with the "expose docker" step which has been flakey and caused build warnings.

/cc @MichaelSimons @weshaggard

### CODE REVIEW

**Comment by MichaelSimons** [comment]
What is the likelihood that this is going to be more stable in the long run?  I have some concerns that we may just be trading issues with this approach.  What are your thoughts on instead of using the "expose docker" step so that we can use the built-in "Publish Artifact" step, that we archive the build logs directly within the Docker container and eliminate the need to copy anything out of the container onto the host?  This is described in https://github.com/dotnet/core-eng/issues/613.

On a side note - your fix would resolve https://github.com/dotnet/core-eng/issues/614.

**Comment by chcosta** [comment]
Sure, there could be some other unforeseen issue that this is trading off with.  I'm pretty sure I've heard you opine that mounted volumes was the generally accepted method for issuing multiple commands to docker, and that stuck with me.  I believe that this is a better approach than the current method because this simplifies things a bit allows direct access to the source code / artifacts, and is more in line with how developers (me) use Docker.  

Archiving artifacts within the docker container is an approach I haven't seriously considered only because it means adding a different tool or process (infrastructure)for this corner case scenario, and that seems to me like it is more likely to cause issues (maintenance) in the long run.  Especially because we would be adding the process solely for the official build scenario.

My preference is to use this method because it fixes the warnings and makes the archive process more reliable (I'd hope). I'm not tightly attached to this approach though and I'm open to alternatives which simplify this world.

**Comment by MichaelSimons** [comment]
Yes mounted volumes are generally accepted and a good pattern to make use of.  The pattern you are proposing I would call `Mount a host directory as a data volume`.  The downside to this pattern is that you often encounter file permission issues which you are working around (chmod).  Core-Setup and CLI use this pattern as well.  IIRC, they workaround it by changing the user running inside the container to match the host user's id.  The pattern I mentioned is referred to is called 'Mount a shared-storage volume as a data volume`.  This pattern is nice as you don't have these ownership issues and it is very easy to cleanup everything as the end of the build by simply deleting the Docker volume.  I completely agree with you that this pattern would require a different tool or process to archive the logs which I am not crazy about.  If we ever transition off of VSTS builds and rely on Jenkins this pattern may be more appropriate.

In the end I don't have an objection with your approach with the caveat that it does have the possibility of causing other flakiness.

File: buildpipeline/DotNet-CoreFx-Trusted-Linux.json [hunk 1]
```diff
@@ -173,7 +138,7 @@
       },
       "inputs": {
         "filename": "docker",
-        "arguments": "exec $(PB_DockerContainerName) $(PB_GitDirectory)/sync.sh $(PB_SyncArguments)",
+        "arguments": "run --rm $(PB_DockerCommonRunArgs) $(PB_DockerVolumeName)/sync.sh $(PB_SyncArguments)",
```
Comment (MichaelSimons) [inline buildpipeline/DotNet-CoreFx-Trusted-Linux.json]: Should --rm be part of PB_DockerCommonRunArgs?

Comment (chcosta) [inline buildpipeline/DotNet-CoreFx-Trusted-Linux.json]: yes, thanks

File: buildpipeline/DotNet-CoreFx-Trusted-Linux.json [hunk 2]
```diff
@@ -393,80 +325,96 @@
...
+    },
+    "PB_CloudResultsAccountName": {
+      "value": "dotnetjobresult"
+    },
+    "PB_CreateHelixArguments": {
+      "value": "\"/p:ArchGroup=x64 /p:ConfigurationGroup=Release /p:TestProduct=corefx /p:TimeoutInSeconds=1200 /p:TargetOS=Linux",
+      "allowOverride": true
+    },
+    "PB_DockerCommonRunArgs": {
+      "value": "--name $(PB_DockerContainerName) -v \"$(PB_GitDirectory):$(PB_DockerVolumeName)\" -w=\"$(PB_DockerVolumeName)\" $(PB_DockerImageName)"
```
Comment (weshaggard) [inline buildpipeline/DotNet-CoreFx-Trusted-Linux.json]: Do you know what using this mapped volume does for performance? I know when I do it locally there is a noticeable slowdown in access the files.

Comment (chcosta) [inline buildpipeline/DotNet-CoreFx-Trusted-Linux.json]: I didn't test that explicitly, run times for the Linux builds appear to be appx on par - https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build?_a=summary&buildId=855235

I neglected to update Linux-Crossbuild with this change, I'm rectifying (validating) that change right now.

**Pull request merged**